File: src/update.js
'use strict';

const { version: PACKAGE_VERSION } = require('../package.json');

const TAGS_API_URL = 'https://api.github.com/repos/kaven-universe/kaven-dns/tags?per_page=100';
const REPOSITORY_URL = 'https://github.com/kaven-universe/kaven-dns';
const DEFAULT_TIMEOUT_MS = 5000;

function parseStableVersion(value) {
  const match = String(value || '').trim().match(/^v?(\d+)\.(\d+)\.(\d+)$/);
  if (!match) return null;
  return {
    version: `${Number(match[1])}.${Number(match[2])}.${Number(match[3])}`,
    parts: match.slice(1).map(Number),
  };
}

function compareVersions(left, right) {
  const a = parseStableVersion(left);
  const b = parseStableVersion(right);
  if (!a || !b) throw new TypeError('Versions must use major.minor.patch format');
  for (let i = 0; i < a.parts.length; i++) {
    if (a.parts[i] !== b.parts[i]) return a.parts[i] > b.parts[i] ? 1 : -1;
  }
  return 0;
}

async function checkForUpdates({
  fetchImpl = globalThis.fetch,
  currentVersion = PACKAGE_VERSION,
  timeoutMs = DEFAULT_TIMEOUT_MS,
} = {}) {
  const current = parseStableVersion(currentVersion);
  if (!current) throw new Error(`Invalid current version: ${currentVersion}`);
  if (typeof fetchImpl !== 'function') throw new Error('Fetch is not available');

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  if (typeof timer.unref === 'function') timer.unref();

  try {
    const response = await fetchImpl(TAGS_API_URL, {
      headers: {
        Accept: 'application/vnd.github+json',
        'User-Agent': `kaven-dns/${current.version}`,
        'X-GitHub-Api-Version': '2022-11-28',
      },
      signal: controller.signal,
    });
    if (!response.ok) throw new Error(`GitHub returned HTTP ${response.status}`);

    const tags = await response.json();
    if (!Array.isArray(tags)) throw new Error('GitHub returned an invalid tag list');

    let latest = null;
    for (const tag of tags) {
      const parsed = parseStableVersion(tag && tag.name);
      if (parsed && (!latest || compareVersions(parsed.version, latest.version) > 0)) {
        latest = { ...parsed, tag: String(tag.name).trim() };
      }
    }
    if (!latest) throw new Error('GitHub returned no stable version tags');

    return {
      currentVersion: current.version,
      latestVersion: latest.version,
      updateAvailable: compareVersions(latest.version, current.version) > 0,
      url: `${REPOSITORY_URL}/tree/${encodeURIComponent(latest.tag)}`,
    };
  } catch (error) {
    if (error && error.name === 'AbortError') throw new Error('Update check timed out');
    throw error;
  } finally {
    clearTimeout(timer);
  }
}

module.exports = { checkForUpdates, compareVersions };

File: internal/update/update.go
package update

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"time"
)

const tagsURL = "https://api.github.com/repos/kaven-universe/kaven-dns/tags?per_page=100"

type Result struct {
	CurrentVersion  string `json:"currentVersion"`
	LatestVersion   string `json:"latestVersion"`
	UpdateAvailable bool   `json:"updateAvailable"`
	URL             string `json:"url"`
}
type Doer interface {
	Do(*http.Request) (*http.Response, error)
}

var stable = regexp.MustCompile(`^v?(\d+)\.(\d+)\.(\d+)$`)

func ParseStable(value string) (string, [3]int, bool) {
	match := stable.FindStringSubmatch(strings.TrimSpace(value))
	if match == nil {
		return "", [3]int{}, false
	}
	var parts [3]int
	for i := range 3 {
		parts[i], _ = strconv.Atoi(match[i+1])
	}
	return fmt.Sprintf("%d.%d.%d", parts[0], parts[1], parts[2]), parts, true
}
func Compare(left, right string) (int, error) {
	_, a, ok := ParseStable(left)
	if !ok {
		return 0, fmt.Errorf("invalid version %q", left)
	}
	_, b, ok := ParseStable(right)
	if !ok {
		return 0, fmt.Errorf("invalid version %q", right)
	}
	for i := range 3 {
		if a[i] > b[i] {
			return 1, nil
		}
		if a[i] < b[i] {
			return -1, nil
		}
	}
	return 0, nil
}
func Check(ctx context.Context, client Doer, currentVersion string) (Result, error) {
	current, _, ok := ParseStable(currentVersion)
	if !ok {
		return Result{}, fmt.Errorf("invalid current version")
	}
	if client == nil {
		client = &http.Client{Timeout: 5 * time.Second}
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, tagsURL, nil)
	if err != nil {
		return Result{}, err
	}
	request.Header.Set("Accept", "application/vnd.github+json")
	request.Header.Set("User-Agent", "kaven-dns/"+current)
	request.Header.Set("X-GitHub-Api-Version", "2022-11-28")
	response, err := client.Do(request)
	if err != nil {
		return Result{}, err
	}
	defer response.Body.Close()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return Result{}, fmt.Errorf("GitHub returned HTTP %d", response.StatusCode)
	}
	var tags []struct {
		Name string `json:"name"`
	}
	decoder := json.NewDecoder(io.LimitReader(response.Body, 1<<20))
	if err := decoder.Decode(&tags); err != nil {
		return Result{}, fmt.Errorf("decode tags: %w", err)
	}
	latest, latestTag := "", ""
	for _, tag := range tags {
		version, _, valid := ParseStable(tag.Name)
		if !valid {
			continue
		}
		if latest == "" {
			latest, latestTag = version, tag.Name
			continue
		}
		comparison, _ := Compare(version, latest)
		if comparison > 0 {
			latest, latestTag = version, tag.Name
		}
	}
	if latest == "" {
		return Result{}, fmt.Errorf("GitHub returned no stable version tags")
	}
	comparison, _ := Compare(latest, current)
	return Result{CurrentVersion: current, LatestVersion: latest, UpdateAvailable: comparison > 0, URL: "https://github.com/kaven-universe/kaven-dns/tree/" + url.PathEscape(strings.TrimSpace(latestTag))}, nil
}
